### PR TITLE
DAOS-11587 vos: standalone vos support for metadata on ssd

### PR DESCRIPTION
**Pull request by wangshilong**
1. extend to support stand vos support for metadata on ssd
2. integration WAL parts with bio: 2.1: format meta header after blobs creation. 2.2: load meta head before open wal and data blob to check if data blobstore is associated and wal blobid.

Required-githooks: true
Signed-off-by: Wang Shilong <shilong.wang@intel.com>

### CODE REVIEW

File: src/bio/bio_xstream.c [hunk 1]
```diff
@@ -1571,7 +1572,9 @@ bio_xsctxt_alloc(struct bio_xs_context **pctxt, int tgt_id, bool self_polling)
 		/* No Data blobstore for sys xstream */
 		if (st == SMD_DEV_TYPE_DATA && tgt_id == BIO_SYS_TGT_ID)
 			continue;
-
+		/* Data blobstore could be only setup after smd prepared */
```
Comment (NiuYawei) [inline src/bio/bio_xstream.c:1575]: I think the special check for sys xstream and standalone VOS are done in the find_xs_bio_dev() already, right?

File: src/bio/bio_xstream.c [hunk 2]
```diff
@@ -1186,7 +1186,8 @@ find_xs_bio_dev(struct bio_xs_context *ctxt, int tgt_id, enum smd_dev_type st,
 	 * Lookup device mapped to @tgt_id and @st in the per-server
 	 * metadata, if found, create blobstore on the mapped device.
 	 */
-	if (tgt_id != BIO_SYS_TGT_ID) {
+	if (tgt_id != BIO_SYS_TGT_ID ||
+	    (tgt_id == BIO_STAND_ALONE_ID && st == SMD_DEV_TYPE_DATA)) {
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/bio/bio_xstream.c:1190]: yes.

File: src/bio/bio_context.c [hunk 3]
```diff
@@ -569,13 +572,67 @@ int bio_mc_create(struct bio_xs_context *xs_ctxt, uuid_t pool_id, uint64_t meta_
...
+	/* fill meta_fmt_info */
+	bxb = bio_xs_context2xs_blobstore(xs_ctxt, SMD_DEV_TYPE_META);
+	uuid_copy(fi->fi_meta_devid, bxb->bxb_blobstore->bb_dev->bb_uuid);
+	bxb = bio_xs_context2xs_blobstore(xs_ctxt, SMD_DEV_TYPE_WAL);
+	uuid_copy(fi->fi_wal_devid, bxb->bxb_blobstore->bb_dev->bb_uuid);
+	/* No data blobstore is possible */
+	if (bxb != NULL) {
+		bxb = bio_xs_context2xs_blobstore(xs_ctxt, SMD_DEV_TYPE_DATA);
+		uuid_copy(fi->fi_data_devid, bxb->bxb_blobstore->bb_dev->bb_uuid);
+	}
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/bio/bio_context.c:602]: uuid_clear

File: src/bio/bio_context.c [hunk 4]
```diff
@@ -569,13 +572,67 @@ int bio_mc_create(struct bio_xs_context *xs_ctxt, uuid_t pool_id, uint64_t meta_
...
+	D_ALLOC_PTR(fi);
+	if (fi == NULL) {
+		rc = -DER_NOMEM;
+		goto delete_wal;
+	}
+
+	if (data_sz == 0)
+		flags |= BIO_MC_FL_NO_DATABLOB;
+	flags |= BIO_MC_FL_NO_CHECKMETA;
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/bio/bio_context.c:588]: OK

File: src/bio/bio_context.c [hunk 5]
```diff
@@ -894,15 +973,17 @@ int bio_mc_close(struct bio_meta_context *bio_mc, enum bio_mc_flags flags)
 		if (rc1)
 			rc = rc1;
 	}
-	if (bio_mc->mc_meta) {
-		rc1 = bio_ioctxt_close(bio_mc->mc_meta, flags);
+	if (bio_mc->mc_wal) {
+		rc1 = bio_ioctxt_close(bio_mc->mc_wal, flags);
 		if (rc1 && !rc)
 			rc = rc1;
+		wal_close(bio_mc);
```
Comment (NiuYawei) [inline src/bio/bio_context.c:980]: wal_close() requires io context, it needs be called before closing wal io context. The closing order should be: wal_close(), meta_close(), wal & meta io context close.

**Review by NiuYawei — CHANGES_REQUESTED** [review]

File: src/bio/bio_context.c [hunk 6]
```diff
@@ -569,13 +617,78 @@ int bio_mc_create(struct bio_xs_context *xs_ctxt, uuid_t pool_id, uint64_t meta_
...
+	if (rc == 0)
+		rc = rc1;
+
+close_meta:
+	rc1 = bio_ioctxt_close(mc->mc_meta, flags);
+	if (rc == 0)
+		rc = rc1;
+delete_wal:
+	if (mc)
+		D_FREE(mc);
```
Comment (jolivier23) [inline src/bio/bio_context.c:679]: D_FREE does this check

File: src/bio/bio_context.c [hunk 7]
```diff
@@ -569,13 +617,78 @@ int bio_mc_create(struct bio_xs_context *xs_ctxt, uuid_t pool_id, uint64_t meta_
...
+	if (fi)
```
Comment (jolivier23) [inline src/bio/bio_context.c:680]: same

**Review by NiuYawei — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**